File: JS/select.js
/**
 * Selection sort implementation in Javascript.
 * Generates and sorts a list of random integers.
 * Tracks and displays # of array accesses and comparisons completed during sorting.
 * 
 * Author: Lucas Burns
 * Version: 2017-9-26
 * 
 */

const process = require('./process.js');

/**
 * Sorts a given array of numbers using Selection sort
 * @param {Array<number>} items 
 */
function sort(items){
	for (var i = 0; i < items.length - 1; i++){
		var min = i;
		for (var j = i; j < items.length; j++){
			if (items[j] < items[min]){
				min = j;
			}
			accesses += 2;
			comparisons++;
		}
		if (min != i){
			process.swap(i, min, items);
		}
	}
}

var a = process.populate();

console.log("Original array:		[" + a + "]");
comparisons = 0;
accesses = 0;

sort(a);

console.log("Sorted array:	 	[" + a + "]");
console.log("%d comparisons		%d array accesses.", comparisons, accesses);

File: python/select.py
#!/usr/bin/env python

"""select.py: Sorts an array of integers using selection sort"""

__author__ = "Lucas Burns"
__version__ = "2017-9-28"

import process

def sort(items):
	global comparisons
	comparisons = 0
	global accesses
	accesses = 0

	for i in range(0, len(items)):
		low = i
		for j in range(i, len(items)):
			if items[j] < items[low]:
				low = j
			accesses += 2
			comparisons += 1
		if low != i:
			process.swap(i, low, items)
			accesses += 3

a = process.populate()

print("Original array:		" + str(a))

sort(a)

print("Sorted array: 		" + str(a))
print(str(comparisons) + " comparisons		" + str(accesses) + " array accesses")